### PR TITLE
[Android] Support React instances with no Activity

### PR DESCRIPTION
**Pull request by igrayson**
This change aims to support applications which run React Native's catalyst instance in the background, sometimes without an `Activity`.

We're using this change today in an RN 0.27 application, which overrides `ReactActivity#createReactInstanceManager` to obtain control of the instance's lifecycle, so that we may run and launch without an `Activity` (i.e., in response to push notifications and other triggers).

We call `CodePush. setReactInstanceManager()` after building our instance (foreground or background).

If we agree with this approach, I'll update this PR with documentation/etc.

### CODE REVIEW

**Comment by geof90** [comment]
Hi @igrayson, overall the approach looks sound! Thanks for your contribution. I have some comments about the implementation, I will post them later.

File: android/app/src/main/java/com/microsoft/codepush/react/CodePushNativeModule.java [hunk 1]
```diff
@@ -99,39 +104,46 @@ public void run() {
...
-                Class instanceManagerHolderClass = currentActivity instanceof ReactActivity
-                        ? ReactActivity.class
-                        : currentActivity.getClass();
-                Field instanceManagerField = instanceManagerHolderClass.getDeclaredField("mReactInstanceManager");
-                instanceManagerField.setAccessible(true);
-                instanceManager = (ReactInstanceManager)instanceManagerField.get(currentActivity);
+            if (instanceManager == null) {
+                final Activity currentActivity = getCurrentActivity();
+
+                if (!ReactActivity.class.isInstance(currentActivity)) {
```
Comment (geof90) [inline android/app/src/main/java/com/microsoft/codepush/react/CodePushNativeModule.java]: This is a regression of https://github.com/Microsoft/react-native-code-push/pull/470. Could you take a look at that and fix this? Basically we no longer want to immediately fallback to the legacy logic the moment the activity is not a ReactActivity.

File: android/app/src/main/java/com/microsoft/codepush/react/CodePush.java [hunk 2]
```diff
@@ -277,6 +278,14 @@ public void clearUpdates() {
         mSettingsManager.removeFailedUpdates();
     }
 
+    public static void setReactInstanceManager(ReactInstanceManager reactInstanceManager) {
+        mReactInstanceManager = reactInstanceManager;
+    }
+
+    static ReactInstanceManager getReactInstanceManager() {
```
Comment (geof90) [inline android/app/src/main/java/com/microsoft/codepush/react/CodePush.java]: private?

File: android/app/src/main/java/com/microsoft/codepush/react/CodePushNativeModule.java [hunk 3]
```diff
@@ -99,39 +104,46 @@ public void run() {
...
+                    // Our preferred reload logic relies on the user's Activity inheriting
+                    // from the core ReactActivity class, so if it doesn't, we fallback
+                    // early to our legacy behavior.
+                    loadBundleLegacy(currentActivity);
+                    return;
+                } else {
+                        ReactActivity reactActivity = (ReactActivity) currentActivity;
```
Comment (geof90) [inline android/app/src/main/java/com/microsoft/codepush/react/CodePushNativeModule.java]: Again a regression of #470, we shouldn't have to cast this.

File: android/app/src/main/java/com/microsoft/codepush/react/CodePushNativeModule.java [hunk 4]
```diff
@@ -87,6 +87,11 @@ private boolean isReactApplication(Context context) {
     }
 
     private void loadBundleLegacy(final Activity currentActivity) {
+        if (currentActivity == null) {
```
Comment (geof90) [inline android/app/src/main/java/com/microsoft/codepush/react/CodePushNativeModule.java]: Since with your change we always pass `getCurrentActivity()` to this method.. I think it would be cleaner to just have `loadBundleLegacy` not take any method params, but simply call `getCurrentActivity()` here and check for nulls here.

File: android/app/src/main/java/com/microsoft/codepush/react/CodePush.java (hunk 2 above)
Comment (igrayson) [inline android/app/src/main/java/com/microsoft/codepush/react/CodePush.java]: It's used by `CodePushNativeModule`

**Comment by igrayson** [comment]
Fixed regressions, caused by basing off of an older `master`.

**Comment by igrayson** [comment]
Updated, updated `README`, and moved to a `InstanceHolder` abstraction.

File: .gitignore [hunk 5]
```diff
@@ -132,6 +132,7 @@ gen/
 .gradle/
 build/
 */build/
+android/app/gradle*
```
Comment (geof90) [inline .gitignore]: Why was this addition needed?

Comment (igrayson) [inline .gitignore]: It may not be. The only way I could wrestle android studio into being able
to navigate/compile the project was to allow it to install a gradle wrapper
at `android/app`.

Separately, it would be great if we had developer setup instructions for
Android Studio, in this repo.

Comment (geof90) [inline .gitignore]: Ah. Normally the way I do development with is to open the demo app project, instead of the library project itself in Android Studio. The demo app project is configured to include the library project as a dependency so those code files would show up.

Comment (geof90) [inline .gitignore]: Could you revert this change then? Or I can remove it after merging.

File: android/app/src/main/java/com/microsoft/codepush/react/CodePushNativeModule.java [hunk 6]
```diff
@@ -143,27 +114,60 @@ private void loadBundle() {
...
         } catch (Exception e) {
             // Our reflection logic failed somewhere
-            // so fall back to restarting the Activity
-            loadBundleLegacy(currentActivity);
+            // so fall back to restarting the Activity (if it exists)
+            loadBundleLegacy();
+        }
+    }
+    
+    private ReactInstanceManager resolveInstanceManager() throws NoSuchFieldException, IllegalAccessException {
```
Comment (geof90) [inline android/app/src/main/java/com/microsoft/codepush/react/CodePushNativeModule.java]: I like this refactoring 👍

**Comment by geof90** [comment]
Thanks @igrayson!
